File: lib/ch1/1.7.js
// Write an algorithm such that if an element in an MxN matrix is 0,
// its entire row and column is set to 0.

const zerofy = matrix => {
  const m = matrix.length;
  const n = matrix[0].length;
  const zeroRows = new Set();
  const zeroCols = new Set();

  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (matrix[i][j] === 0) {
        zeroRows.add(i);
        zeroCols.add(j);
      }
    }
  }

  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (zeroRows.has(i) || zeroCols.has(j)) {
        matrix[i][j] = 0;
      }
    }
  }

  return matrix;
};

module.exports = zerofy;

File: java/ch1/1.7.java
public static void zerofy(int[][] matrix) {
    int m = matrix.length;
    int n = matrix[0].length;
    Set<Integer> zeroRows = new HashSet<>();
    Set<Integer> zeroCols = new HashSet<>();
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (matrix[i][j] == 0) {
                zeroRows.add(i);
                zeroCols.add(j);
            }
        }
    }
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (zeroRows.contains(i) || zeroCols.contains(j)) matrix[i][j] = 0;
        }
    }
    System.out.println(matrix);
}
